### PR TITLE
Allow to change the representation style of null values

### PR DESCRIPTION
**Pull request by frodrigo**
The null values are by default represented according to the type of map :

* by a black polygon
* or by a grey circle of fixed radius

You must be able to set a default style when no value is set:

* in the case of a polygon: fill color, outline color
* in the case of a circle: circle size, fill color, contour color

Proposal:

* For circles, by default no poster
* For polygons only the contour

Possibility to override the default, to change the style or not to display.
If the feature is displayed, then there is a caption that goes with it.

### CODE REVIEW

File: terra_layer/style.py [hunk 1]
```diff
@@ -20,24 +26,40 @@
     },
 }
 
+DEFAULT_LEGENDE_GRADUADED = {
```
Comment (jrmi) [inline terra_layer/style.py:29]: `LEGEND` without e ?

File: terra_layer/style.py [hunk 2]
```diff
@@ -455,31 +554,58 @@ def generate_style_from_wizard(layer, config):
...
+                "items": gen_legend_steps(
+                    boundaries, colors, include_no_value, no_value_fill_color
+                )[::-1],
             }
             return (style, legend_addition)
         else:
-            return (DEFAULT_STYLE_GRADUADED, {})
+            return (
+                DEFAULT_STYLE_GRADUADED,
+                {"items": [DEFAULT_LEGENDE_GRADUADED]} if include_no_value else {},
```
Comment (jrmi) [inline terra_layer/style.py:588]: Without E

File: terra_layer/style.py [hunk 3]
```diff
@@ -488,28 +614,45 @@ def generate_style_from_wizard(layer, config):
...
+                    mm[0], mm[1], config["max_diameter"], fill_color, include_no_value,
                 ),
                 "stackedCircles": True,
             }
             return (style, legend_addition)
         else:
-            return (DEFAULT_STYLE_CIRCLE, {})
+            return (
+                DEFAULT_STYLE_CIRCLE,
+                {"items": [DEFAULT_LEGENDE_CIRCLE]} if include_no_value else {},
```
Comment (jrmi) [inline terra_layer/style.py:654]: Without e

File: terra_layer/style.py [hunk 4]
```diff
@@ -455,31 +554,58 @@ def generate_style_from_wizard(layer, config):
...
         """ config = {
             "field": "my_field",
             "symbology": "circle",
+            "include_no_value": False,  # Show no value features on map and legend
             "max_diameter": 200,
             "fill_color": "#0000cc",
             "fill_opacity": 0.5,
             "stroke_color": "#ffffff",
             "stroke_width": 1,
+            "no_value_fill_color": "#000000",
```
Comment (jrmi) [inline terra_layer/style.py:601]: can you use a dict like:

```python
{
  "no_value_style" : {
    "fill_color": …
  }
}
```

File: terra_layer/style.py [hunk 5]
```diff
@@ -455,31 +554,58 @@ def generate_style_from_wizard(layer, config):
                 'With "graduated" symbology, "boundaries" or "method" should be provided'
             )
 
+        include_no_value = config.get("include_no_value", True)
         if boundaries:
+            field_style = get_field_style(field)
+            no_value_fill_color = config.get(
```
Comment (jrmi) [inline terra_layer/style.py:560]: default_style should be a dict that you can merge with default values.

Comment (frodrigo) [reply terra_layer/style.py]: Ok, use a dict with default values.

File: terra_layer/style.py (hunk 4 above)
Comment (frodrigo) [reply terra_layer/style.py:607]: Keep it flat to use with default values.

Comment (jrmi) [reply terra_layer/style.py:607]: Don't see the point...

File: terra_layer/style.py [hunk 6]
```diff
@@ -386,27 +452,49 @@ def gen_legend_circle(min, max, size, color):
         for b in boundaries
     ]
 
+    if include_no_value:
+        ret += [
```
Comment (jrmi) [inline terra_layer/style.py:456]: `ret.append` instead ?

Comment (frodrigo) [reply terra_layer/style.py:456]: As you wish, Nahuell tell me .append() is too java-ish for python.

Comment (frodrigo) [reply terra_layer/style.py]: Fixed.

File: terra_layer/style.py [hunk 7]
```diff
@@ -455,31 +569,60 @@ def generate_style_from_wizard(layer, config):
...
         """ config = {
             "field": "my_field",
             "symbology": "circle",
+            "include_no_value": False,  # Show no value features on map and legend
             "max_diameter": 200,
             "fill_color": "#0000cc",
             "fill_opacity": 0.5,
             "stroke_color": "#ffffff",
             "stroke_width": 1,
+            "no_value_fill_color": "#000000",
```
Comment (jrmi) [inline terra_layer/style.py:618]: You should update the doc.

File: terra_layer/style.py [hunk 8]
```diff
@@ -489,27 +632,45 @@ def generate_style_from_wizard(layer, config):
             radius_base = ["sqrt", ["/", get_field_style(field), ["pi"]]]
             radius = gen_style_interpolate(radius_base, boundaries, sizes)
 
+            config_style = {
```
Comment (jrmi) [inline terra_layer/style.py:635]: Why don't you use update ?

Comment (frodrigo) [reply terra_layer/style.py:635]: It is more concise to do like this, than clone, update and filter the dict.

File: terra_layer/style.py (hunk 7 above)
Comment (frodrigo) [reply terra_layer/style.py:618]: Do we have a doc?

File: terra_layer/style.py [hunk 9]
```diff
@@ -365,7 +423,15 @@ def boundaries_round(boundaries, scale=2):
...
+    max,
+    size,
+    color,
+    include_no_value,
+    no_value_circle_radius=DEFAULT_NO_VALUE_CIRCLE_RADIUS,
+    no_value_color=DEFAULT_NO_VALUE_FILL_COLOR,
+):
     """
     Generate a circle legend.
     """
```
Comment (jrmi) [inline terra_layer/style.py:437]: You can't set dict as default value directly.

```suggestion
def gen_legend_circle(
    min,
    max,
    size,
    color,
    include_no_value,
    no_value_circle_radius=DEFAULT_NO_VALUE_CIRCLE_RADIUS,
    no_value_color=None,
):
    """
    Generate a circle legend.
    """
    no_value_color= no_value_color or dict(DEFAULT_NO_VALUE_FILL_COLOR)
```

File: terra_layer/style.py [hunk 10]
```diff
@@ -437,10 +547,14 @@ def generate_style_from_wizard(layer, config):
              "field": "my_field",
              "symbology": "graduated",
              "boundaries": [1, 2, 3, 5],
+             "include_no_value": True,  # Show no value features on map and legend
              "method": "equal_interval",  # How to compute boundaries if not provided
              "fill_color": ["#ff0000", "#aa0000", "#770000", "#330000", "#000000"],
              "fill_opacity": 0.5,
              "stroke_color": "#ffffff",
+             "no_value_fill_color": "#000000",
```
Comment (jrmi) [inline terra_layer/style.py:555]: Should be a dict also

File: terra_layer/style.py (hunk 7 above)
Comment (jrmi) [reply terra_layer/style.py:618]: This comment is a doc.

File: terra_layer/style.py (hunk 9 above)
Comment (jrmi) [reply terra_layer/style.py:437]: May be we can remove default here as their are defined before...

File: terra_layer/style.py [hunk 11]
```diff
@@ -455,31 +569,60 @@ def generate_style_from_wizard(layer, config):
...
             "max_diameter": 200,
             "fill_color": "#0000cc",
             "fill_opacity": 0.5,
             "stroke_color": "#ffffff",
             "stroke_width": 1,
+            "no_value_fill_color": "#000000",
+            "no_value_fill_opacity": 0,
+            "no_value_stroke_color": "#ffffff",
+            "no_value_stroke_width": 1,
         }
```
Comment (jrmi) [inline terra_layer/style.py:622]:
```suggestion
        """ config = {
            "field": "my_field",
            "symbology": "circle",
            "include_no_value": False,  # Show no value features on map and legend
            "max_diameter": 200,
            "style": {
                "fill_color": "#0000cc",
                "fill_opacity": 0.5,
                "stroke_color": "#ffffff",
                "stroke_width": 1,
            },
            no_value_style{
                 "fill_color": "#000000",
                "fill_opacity": 0,
                "stroke_color": "#ffffff",
                "stroke_width": 1,
            }
            

        }
```

Comment (frodrigo) [reply terra_layer/style.py]: OK

File: terra_layer/style.py (hunk 10 above)
Comment (frodrigo) [reply terra_layer/style.py]: OK

File: terra_layer/style.py (hunk 9 above)
Comment (frodrigo) [reply terra_layer/style.py]: OK

File: terra_layer/style.py (hunk 7 above)
Comment (frodrigo) [reply terra_layer/style.py]: OK

**Pull request merged**